### PR TITLE
[metadata] integration metadata collection

### LINKED ISSUE #816: Collect application metadata

### PR DESCRIPTION
**Pull request by yannmh**
* Interface for collecting metadata
* Splitting payload to distinguish data and metadata
* Two endpoints: /intake/metrics and /intake/metadata
* '_collect_metadata' functions for a few checks
* Tests

Rebased on 5.2.1 agent release
Fixes #816

### CODE REVIEW

File: checks/collector.py [hunk 1]
```diff
@@ -19,16 +20,96 @@
...
+    """
+    METADATA_KEYS = frozenset(['meta', 'tags', 'host-tags', 'systemStats',
+                               'agent_checks', 'gohai', 'external_host_tags'])
+
+    DUPLICATE_KEYS = frozenset(['apiKey', 'agentVersion'])
+
+    DATA_ENDPOINT = 'metrics'
+    METADATA_ENDPOINT = 'metadata'
+
+    def __init__(self, start_event=True):
```
Comment (LeoCavaille) [inline checks/collector.py]: I don't really get why we carry around this `start_event` ?

File: ddagent.py [hunk 2]
```diff
@@ -407,7 +424,8 @@ def _postMetrics(self):
 
     def run(self):
         handlers = [
-            (r"/intake/?", AgentInputHandler),
+            (r"/intake/metrics?", MetricsAgentInputHandler),
+            (r"/intake/metadata?", MetadataAgentInputHandler),
```
Comment (LeoCavaille) [inline ddagent.py]: Mmh, should we use different endpoints like that, our use different base URLs/ELBs ? or both?
cc @remh 

We need to be in sync for that one.

File: ddagent.py [hunk 3]
```diff
@@ -211,8 +214,10 @@ def get_url(self, endpoint):
         endpoint_base_url = self.get_url_endpoint(endpoint)
         api_key = self._application._agentConfig.get('api_key')
         if api_key:
-            return endpoint_base_url + '/intake?api_key=%s' % api_key
-        return endpoint_base_url + '/intake'
+            return endpoint_base_url + '/intake/%s'\
+                '?api_key=%s' % (self._msg_type, api_key)
+        return endpoint_base_url + '/intake/%s'\
+            % self._msg_type
```
Comment (LeoCavaille) [inline ddagent.py]: Could you use a format instead, I find them much more readable?

File: tests/test_common.py [hunk 4]
```diff
@@ -264,5 +265,86 @@ def test_dupe_tags(self):
...
+            if metadata:
+                self._collect_metadata()
+
+        def _collect_metadata(self):
+            self.svc_metadata({'version': 1})
+
+    def test_metadata_length(self):
+        # Fake config. Instance 2 will not generate any meta
+        config = {'instances': [{'metadata': True}, {'metadata': False}]}
+        agentConfig = {
```
Comment (LeoCavaille) [inline tests/test_common.py]: `{}`

File: tests/test_common.py [hunk 5]
```diff
@@ -264,5 +265,86 @@ def test_dupe_tags(self):
         metric = self.aggr.metrics.values()[0]
         self.assertEquals(metric.value, 2)
 
+
+class TestMetadataService(unittest.TestCase):
```
Comment (LeoCavaille) [inline tests/test_common.py]: Can we put that in a different test file test_metadata or something?

File: tests/test_common.py [hunk 6]
```diff
@@ -264,5 +265,86 @@ def test_dupe_tags(self):
...
+        check.run()
+        service_metadata = check.get_service_metadata()
+        service_metadata_count = len(service_metadata)
+
+        self.assertEquals(service_metadata_count, 2)
+        self.assertEquals(service_metadata[0], {'version': 1})
+        self.assertEquals(service_metadata[1], {})
+
+
+class TestAgentPayload(unittest.TestCase):
```
Comment (LeoCavaille) [inline tests/test_common.py]: Same here, I'd like to avoid this 300 line file.

**Comment by LeoCavaille** [comment]
Also I didn't see any way to easily display the metadata locally, perhaps you can add something to the `info` display with the collector status and/or to the `dd-agent check` command?

**Comment by LeoCavaille** [comment]
Looks good, and I like the idea of refactoring a bit the payload and emitter part, I'll try to spend more time to see if we can do that more in depth and simplify the existing code.

**Comment by LeoCavaille** [comment]
Also looks like the failures on Travis are legit! If you can take a look.

**Comment by yannmh** [comment]
Closing in favor of #1611

**Pull request closed without merging**